### PR TITLE
Fix masking issues in K.rnn (tensorflow + theano)

### PR DESCRIPTION
**Pull request by andhus**
### Summary
EDIT: Fixes https://github.com/keras-team/keras/issues/11472 _and_ https://github.com/keras-team/keras/issues/11498 for Theano and Tensorflow backend.

~~Fixes https://github.com/keras-team/keras/issues/11498 for tensorflow backend.~~

~~**NOTE** that this PR is a continuation of https://github.com/keras-team/keras/pull/11473 which solves the related issue: https://github.com/keras-team/keras/issues/11472.~~
 
### PR Overview
- [y] This PR requires new unit tests [y/n] (make sure tests are included)
- [n] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [y] This PR is backwards compatible [y/n]
- [n] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -2897,9 +2897,22 @@ def rnn(step_function, inputs, initial_states,
     if mask is not None:
         if mask.dtype != tf.bool:
             mask = tf.cast(mask, tf.bool)
-        if len(mask.get_shape()) == ndim - 1:
-            mask = expand_dims(mask)
-        mask = tf.transpose(mask, axes)
```
Comment (andhus) [inline keras/backend/tensorflow_backend.py]: This shows that we're on the "wrong track", as the masking shape should not depend on input (which `axes` are derived from)

File: keras/backend/tensorflow_backend.py [hunk 2]
```diff
@@ -2897,9 +2897,22 @@ def rnn(step_function, inputs, initial_states,
...
+        if len(mask.get_shape()) != 2:
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: You can use `.shape` nowadays

File: keras/backend/tensorflow_backend.py [hunk 3]
```diff
@@ -2897,9 +2897,22 @@ def rnn(step_function, inputs, initial_states,
...
+            raise ValueError(
+                "mask should have shape=(samples, time), "
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Nit: use `'` as the quote character, for consistency

Comment (fchollet) [reply keras/backend/tensorflow_backend.py]: Nit: use backquotes around code keywords

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -2981,26 +2980,29 @@ def rnn(step_function, inputs, initial_states,
...
             size=time_steps,
             tensor_array_name='output_ta')
+        initial_output = zeros_like(output)
         input_ta = tensor_array_ops.TensorArray(
             dtype=inputs.dtype,
             size=time_steps,
             tensor_array_name='input_ta')
         input_ta = input_ta.unstack(inputs)
         time = tf.constant(0, dtype='int32', name='time')
+        while_loop_kwargs = dict(
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: Nit: use 
```python
while_loop_kwargs = {
   'cond': lambda time, *_: time < time_steps,
   ...
}
```

File: tests/keras/backend/backend_test.py [hunk 5]
```diff
@@ -970,6 +970,60 @@ def step_function(x, states):
         assert_list_pairwise(last_output_list, shape=False)
         assert_list_pairwise(outputs_list, shape=False)
 
+    def test_rnn_output_ndim_larger_than_2_masking(self):
+        n_samples = 3
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Nit: use `num_` for counters, for consistency

**Comment by andhus** [comment]
Thanks for the review @fchollet. I've now fixed the issues in Theano backend as well (see https://github.com/keras-team/keras/pull/11499/commits/698543524558bf5e2f72c927c1a4f4f211652695). Status for CNTK is a bit unclear, new tests as disabled for now.

**Comment by gabrieldemarmiesse** [comment]
@andhus could you merge master into your branch to resolve the conflicts? It will also remove pytest-timeout and hopefully help with the travis build.

File: tests/keras/layers/recurrent_test.py [hunk 6]
```diff
@@ -165,6 +165,90 @@ def test_masking_correctness(layer_class):
...
+            super(Cell, self).__init__()
+
+        def build(self, input_shape):
+            self.state_size = input_shape[-1]
+            self.output_size = input_shape[-1]
+
+        def call(self, inputs, states):
+            return inputs, [s + 1 for s in states]
+
+    for unroll in [True, False]:
```
Comment (gabrieldemarmiesse) [inline tests/keras/layers/recurrent_test.py]: I think you can also here test with random values, but check that the numpy arrays given by `model.predict` are the same   for both cases.

File: tests/keras/layers/recurrent_test.py [hunk 7]
```diff
@@ -165,6 +165,90 @@ def test_masking_correctness(layer_class):
...
+            super(Cell, self).__init__()
+
+        def build(self, input_shape):
+            self.state_size = input_shape[-1]
+            self.output_size = input_shape[-1] * 2
+
+        def call(self, inputs, states):
+            return keras.layers.concatenate([inputs] * 2), [s + 1 for s in states]
+
+    for unroll in [True, False]:
```
Comment (gabrieldemarmiesse) [inline tests/keras/layers/recurrent_test.py]: Same here.

**Comment by gabrieldemarmiesse** [comment]
Just a some comments for the robustness of the tests and ensure that this kind of error doesn't happen again. 

You might need to modify the numpy implementation of `rnn` if the implementation have the same bug. You can find it here: https://github.com/keras-team/keras/blob/0354142fa447e343ff99eda1ed504c19d4474aeb/keras/backend/numpy_backend.py#L188

File: tests/keras/backend/backend_test.py [hunk 8]
```diff
@@ -970,6 +970,107 @@ def step_function(x, states):
...
 
+    def test_rnn_output_and_state_masking_independent(self):
+        num_samples = 2
+        num_timesteps = 4
+        state_and_io_size = 5
+
+        def step_function(inputs, states):
+            return inputs, [s + 1 for s in states]
+
+        inputs_vals = np.ones((num_samples, num_timesteps, state_and_io_size))
```
[the remark this answers is not in the document]
Comment (andhus) [reply tests/keras/backend/backend_test.py]: I think the `WITH_NP` approach makes sense for the "basic tensor operations" where the numpy implementation represents the reference source of truth. However, for more complicated operations (such as the `rnn`) I think it is bad to test the backends by comparing to the reference backend. If we do that  **we must write exactly this kind of tests with "by hand calculation" to verify _its_ functionality** And when we do, we might just as well run those tests  on the real backends?

Infact, I think it is worrying that there aren't proper input/output test for some methods in the numpy backend. Take the `pool` function for example - it is 50 lines of code that we just take as ground truth, but there are no tests for this code veryfying that it does what we want it to!? (Or am I just not finding those tests?)

Allthough, for completness, it would be nice to have a functioning `rnn` implementation in the numpy backend. However, it is curently nowhere near a proper implementation (it has a different signature and deals with the very special case of a "SimpleRNN") so I don't think that's within the scope of this PR.

**Comment by andhus** [comment]
@gabrieldemarmiesse Please see my answer to your review regarding `WITH_NP` reference implementation for `rnn`, it is not "at all" implemented currently. I followed your suggestions and changed all inputs to random numbers and clarified that expected outputs are the same for `unroll=True/False`. There is no need for the Numpy reference implementation for this PR, the tests are self-contained and verify the expected behaviour.

File: tests/keras/backend/backend_test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (gabrieldemarmiesse) [reply tests/keras/backend/backend_test.py]: Actually, what we do right now is, when running travis, we compare:
* theano against numpy
* tensorflow against numpy
* cntk against numpy

those are three different jobs. If one backend is wrong, the corresponding job will fail. If the numpy backend is wrong, the three jobs will fail.

Running some kind of correctness computation is really difficult to do right, because you would need to use random input values to ensure correctness in a majority of cases. But then if you take random values as input, it's very likely that you won't use a paper and a pencil to compute the expected output. You'll likely use numpy. So I think this is our best bet to avoid shallow tests.

After all this, I didn't dig much into the rnn implementation in the numpy backend, so I can't guarantee that it supports your use case. And if it does not, it would surely fall outside the scope of this PR, and we're better off with your implementation of the tests for sure.

After a bit of digging into the git blame, I found that the numpy RNN implemtation was done there: #9557

Maybe @taehoonlee can tell us more about it?

Otherwise I'm fine with the tests that you made if it means changing a lot of code to adapt it to my request. I don't intend to make you do too much extra work since you're already helping a lot with your frequent PRs and discussions :)

**Comment by gabrieldemarmiesse** [comment]
@andhus Thanks for your quick response. I agree that the tests are already good enough as-is. I though we could save some LOCs by using the numpy rnn, but it seems that the API, so it can be postponed. Thanks for your work again!

**Comment by gabrieldemarmiesse** [comment]
I'll let @fchollet decide if/when to merge.

File: tests/keras/backend/backend_test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (andhus) [reply tests/keras/backend/backend_test.py]: Thanks for your detailed response @gabrieldemarmiesse . I definitely see how the reference implementation is convenient and improves test coverage 👍  My main point is that if we _only_ rely on "testing by agreement" then _agreement_ is also the only guarantee we get from the tests.

> those are three different jobs. If one backend is wrong, the corresponding job will fail. If the numpy backend is wrong, the three jobs will fail.

But all jobs will run on the same commit, which means if I make a small modification in _all backends_ in the same commit/PR (which is almost what I do here!) then a backward **in**compatible change can go undetected. I understand that this is not likely to happen in practice, but principally it requires the external procedure that you are not allowed to change all backends simultaneously to be safe.

Even if the backends are separate implementations they are still "logically" strongly coupled which means that it is not "completely unlikely" that a bug could be introduced simultaneously in all backends (think e.g. an "innocent PR" aligning variable naming in all of them for improved readability).

It also makes it harder to reason about the code changes; if you have version _X_ and a later version _Y_  of the code (possibly separated by hundreds of commits) and want to verify if a certain function _f_ in _Y_ is backwards compatible with respect to _X_. With input/output unit tests it is a sufficient guarantee that the tests of _f_ in _X_ are still there in _Y_ (assuming good test coverage). To do the same thing with the "agreement tests" you would basically need the full history of Travis logs and check that there are never simultaneous changes in all backends between registered passing tests...

My bottom line is just that (at least for the "more complicated" backend operations) it is worth having the hand calculated tests - for proper "grounding" 😄

[the remark this answers is not in the document]
Comment (andhus) [reply tests/keras/backend/backend_test.py]: ...It shouldn't be that much work to update the numpy rnn, but let's do it in another PR (and let's keep the hand calculated i/o tests even when its introduced 🙏)

[the remark this answers is not in the document]
Comment (gabrieldemarmiesse) [reply tests/keras/backend/backend_test.py]: Sure thing! I agree that those hand calculated test are important and we'll keep them. You made a very good point.

**Comment by staqiuddin** [comment]
Thank you all for your great efforts. @andhus, I just have a question regarding the issues fixed by this PR. Sorry if the question is trivial but I want to be very sure about this.

I am using the `LSTM` and `GRU` layers with masking, i.e., a `Masking` layer precedes these two layers. I found (in the source code) that these two layers are sub-classes of the `RNN` layer. Therefore, masking in `LSTM` and `GRU` is affected by #11472 and #11498 within the same context/setting (including backend) defined in these two issues. Am I right?

Thank you.

**Comment by andhus** [comment]
Thanks @staqiuddin :)

Good question; It is correct that both `LSTM` and `GRU` are based on `RNN` and thereby `K.rnn` with the mentioned issues. *However*, for both of these recurrent layers the following holds:
1) number of dimensions of  `output` and all `states[i]` are 2
2) `output == states[0]`

Which means that _masking will have worked fine for these layers_ - no need to worry 👍

**Comment by staqiuddin** [comment]
Thank you very much @andhus for your reply. It is _very_ good to hear that. Much appreciated.
